### PR TITLE
Address Windows port binding errors

### PR DESCRIPTION
**Pull request by emabrey**
**This is a copy of #1027 created to transition the merge to the new default branch main.**
**This has been rebased onto main and then cherry-picked.**

On Windows 10 running either WSL or Hyper-V, the operating system will acquire random and ephemeral binds of various port ranges. Once these port ranges are allocated for the usage of these two OS features, they are no longer available for usage by other processes, even though at a given point in time they may technically not be actively bound. It just so happens that on these machines, the default port of sccache (TCP 4226) conflicts with the default excluded port ranges for some configurations of winNAT and Hyper-V. The ideal way to resolve this would be to tell the user to modify the operating system to pick it's TCP ports from much higher in the TCP port space (such as the dynamic TCP port range). However, this is much more complicated than simply manually instructing sccache to utilize a different port than the default, not to mention the user is more likely to blame sccache for the failure than the OS regardless of who is actually in the wrong. For information on how the Docker project handled this same problem, see docker/for-win#3171.

Within sccache, to address this issue the proper method would be usage of the environmental variable `SCCACHE_SERVER_PORT`, however, the documentation does not go into detail on the availability of this configurable. To make users aware of how to fix this issue, I have added a warning message that instructs the user on how to change the default sccache port via the variable. Additionally, instead of hoping that the port we are using will be valid when sccache attempts to bind it, now startup of the sccache executable verifies that the port we will use, default or manually configured, is valid. This is much better than the previous behavior of letting the operating system give the user cryptic and essentially unhelpful error messages (for example OS Error 10013, which I experienced myself and which other people have also experienced e.g. #413).

This fix was implemented using the addition of a new function which runs between log initialization and the handoff to the main commands dispatch logic. This function verifies that either the `commands::DEFAULT_PORT` or the `SCCACHE_SERVER_PORT` values are valid and bindable ports, but only if we are running a `--start-server` command, since checking for the bind on other options would be incorrect, especially in the `--stop-server` scenario. To enable this I needed to make a few modifications to the `cmdline::StatsFormat` and `cmdline::Commands` enums, specifically I needed to add the `PartialEq` trait so that the port check code would be able to tell what the command line args were without resorting to a match statement. If the check fails the `sccache::main` function will quickly exit using `std::process::exit(1)` to indicate an error occurred. The check function can optionally emit the aforementioned error messages to stderr, but in the `sccache::main` usecase it **must** emit the error messages to comply with the unit tests' expectation that a failed `--start-server` will emit the text `Server startup failed:`.

The logic for evaluating if a port is "valid" utilizes a new dependency on `getport-4.0.0` to provide the ability to test the validity of the port binds without resorting to ugly command line parsing.

Resolves mozilla/sccache#413

#### `SCCACHE_SERVER_PORT` set but not valid
image

#### `SCCACHE_SERVER_PORT` not set (using invalid default port)
image

#### `SCCACHE_SERVER_PORT` set using valid value
image

### CODE REVIEW

**Comment by emabrey** [comment]
Actually, it is possible to run `netsh int ipv4 add excludedportrange protocol = tcp startport = (u16) numberofports = 1` to exclude a port manually, so if we coordinated the integration test correctly it could be a valid windows only unit test. Ultimately it would just look for the error message. That can get rid of the TOCTOU problem. We would just need to run `netsh int ipv4 delete excludedportrange protocol = tcp startport = (u16) numberofports = 1` to cleanup afterwards to stop the test from stomping on itself. I'll write a new windows only integration test that tries that and I'll report back if that works.

**Review by mitchhentges — CHANGES_REQUESTED** [review]
> Actually, it is possible to run netsh int ipv4 add excludedportrange protocol = tcp startport = (u16) numberofports = 1 to exclude a port manually

Woah, I agree with you that that's too much! It might even require admin privileges on CI workers, which I'd like to avoid.

Looking into testing `run_command()` and `start_server()`, I don't see a convenient way of injecting mock implementations, so I think I'm happy to land this without tests once it's green and approved.

Thanks again for the patch and continued improvements!

File: src/commands.rs
Comment (emabrey) [reply src/commands.rs:597]: A `SeverStartup::AddrInUse` only happens when the port is bound. The problem with 10013 is that the port is in an excluded range, not that it is or isn't bound. The applications are using [`CreatePersistentTcpPortReservation`](https://docs.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-createpersistenttcpportreservation) to reserve an entire range of TCP ports and only binding to them when they choose. The reservation makes sure nobody else can take your port range, even if the ports were configured for reusable binding. Instead of using the command line I could just call into that function from Rust using FFI. I know winapi is already a dependency, and it includes the `iphlpapi.h` header defines as "iphlpapi"

So with 
```

[target.'cfg(windows)'.dev-dependencies.winapi]
version = "0.3"
features = [
    "iphlpapi"
]
```

I could write a test that uses the appropriate functions and just marks itself as a pass if it gets a `ERROR_ACCESS_DENIED` result (with maybe a warning message about some tests being unable to run without admin permissions).

Comment (mitchhentges) [reply src/commands.rs:597]: > A SeverStartup::AddrInUse only happens when the port is bound.

`ServerStartup` is defined by us in `sccache` itself, so we control which OS errors get mapped to it.
[Currently, `io::ErrorKind::AddrInUse` the only kind of error mapped to `ServerStartup::AddrInUse`](https://github.com/mozilla/sccache/blob/6488b3c7fd7145940d16e52fe1022a9249dad696/src/server.rs#L426), but we can modify this code.

> The applications are using [CreatePersistentTcpPortReservation](https://docs.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-createpersistenttcpportreservation) to reserve an entire range of TCP ports and only binding to them when they choose. The reservation makes sure nobody else can take your port range, even if the ports were configured for reusable binding. Instead of using the command line I could just call into that function from Rust using FFI. I know winapi is already a dependency, and it includes the iphlpapi.h header defines as "iphlpapi"

I don't quite follow you here: I'm not sure if you're referring to the tests or the "is port already in-use/allocated/reserved/not-accessible" behaviour at runtime.
I _don't_ think that we should be calling `[CreatePersistentTcpPortReservation](https://docs.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-createpersistenttcpportreservation), because we only need the one port, not multiple - and we still have to handle the same `10013` error anyways.

> I could write a test that uses the appropriate functions and just marks itself as a pass if it gets a ERROR_ACCESS_DENIED result (with maybe a warning message about some tests being unable to run without admin permissions).

This test would be flaky depending on whether it's running on a system with WSL/Hyper-V, or if it has admin permissions. I don't think it's worth it personally.

File: src/commands.rs
Comment (emabrey) [reply src/commands.rs:605]: Done.

File: src/commands.rs
Comment (emabrey) [reply src/commands.rs]: There is no easy way to implement the errnum check in lower code as far as I can see. To keep the performance benefit I moved the code into Windows specific implementation of a function. I also added a guard statement so that if the String comparison breaks or changes form (which should be impossible unless we start generating non OS error messages) we get a Debug message telling the dev to check the statements for consistency. It shouldn't break though since `Error::os_last_error().into()` generates a `String` the same way that `Error::from_raw_os_error(u16).to_string()` does.

Comment (emabrey) [reply src/commands.rs]: So I looked into what `ErrorKind` is issued for the 10013 errnum, and it's an `ErrorKind::Uncategorized`. As a result I think that taking the `ErrorKind` approach is really not that useful, since we have to rely on unstable Rust.

**Comment by emabrey** [comment]
Can you review this again? I've made significant changes.

File: src/commands.rs
Comment (mitchhentges) [reply src/commands.rs]: > There is no easy way to implement the errnum check in lower code as far as I can see.

I don't understand, why can't we edit the code [where we map from `io::Error` to `ServerStartup`](https://github.com/mozilla/sccache/blob/6488b3c7fd7145940d16e52fe1022a9249dad696/src/server.rs#L426)?
I'm guessing it'd look something like this:

```
match e.downcast_ref::<io::Error>() {
    Some(io_err) if io::ErrorKind::AddrInUse == io_err.kind() => {
        notify_server_startup(&notify, ServerStartup::AddrInUse)?;
    }
    Some(io_err) if cfg!(windows) && Some(10013) == io_err.raw_os_error() => {
        // 10013 is the "WSAEACCES" error, which can occur if the requested port
        // has been allocated for other purposes, such as winNAT or Hyper-V.
        notify_server_startup(&notify, ServerStartup::AddrInUse)?;
    }
    _ => {
        let reason = e.to_string();
        notify_server_startup(&notify, ServerStartup::Err { reason })?;
    }
};
```

Does that make sense?

**Comment by mitchhentges** [comment]
> Can you review this again? I've made significant changes.

For future reference: the review continued [as part of this comment here](https://github.com/mozilla/sccache/pull/1128#discussion_r817164815).

**Comment by emabrey** [comment]
> > Can you review this again? I've made significant changes.
> 
> For future reference: the review continued [as part of this comment here](https://github.com/mozilla/sccache/pull/1128#discussion_r817164815).

Just bumping this to indicate looking at this is on my to-do list, I've just been quite busy.

**Comment by drahnr** [comment]
Gentle ping. Is there some maintainer confirming this works as expected and the approach is sane?

**Comment by emabrey** [comment]
I did not mean to remove this PR. I didn't realize that Github would automatically close it if I rebased the branch the way I did. Whoops. So, I don't have any choice to make a new PR because I don't have the permissions to re-open this one.

**Pull request closed without merging**
